pgmax/fg/groups.py

```diff
@@ -25,104 +22,76 @@
 import pgmax.fg.nodes as nodes
 from pgmax.utils import cached_property
 
+MAX_SIZE = 1e9
 
+
+@total_ordering
 @dataclass(frozen=True, eq=False)
 class VariableGroup:
     """Class to represent a group of variables.
+    Each variable is represented via a tuple of the form (variable hash, variable num_states)
 
-    All variables in the group are assumed to have the same size. Additionally, the
-    variables are indexed by a variable name, and can be retrieved by direct indexing (even indexing
-    a sequence of variable names) of the VariableGroup.
-
-    Attributes:
-        _names_to_variables: A private, immutable mapping from variable names to variables
+    Arguments:
+        num_states: An integer or an array specifying the number of states of the variables
+            in this VariableGroup
     """
 
-    _names_to_variables: Mapping[Hashable, nodes.Variable] = field(init=False)
+    num_states: Union[int, np.ndarray]
 
-    def __post_init__(self) -> None:
-        """Initialize a private, immutable mapping from variable names to variables."""
-        object.__setattr__(
-            self,
-            "_names_to_variables",
-            MappingProxyType(self._get_names_to_variables()),
-        )
-
-    @typing.overload
-    def __getitem__(self, name: Hashable) -> nodes.Variable:
-        """This function is a typing overload and is overwritten by the implemented __getitem__"""
+    def __post_init__(self):
+        # Only compute the hash once, which is guaranteed to be an int64
+        this_id = id(self) % 2**32
+        this_hash = this_id * int(MAX_SIZE)
+        assert this_hash < 2**63
+        object.__setattr__(self, "this_hash", this_hash)
```

Comment:
Can we change `this_hash` to something like `_hash` to emphasize that this shouldn't be publicly accessible?